geode-core/src/main/java/org/apache/geode/internal/cache/TXRegionLockRequestImpl.java

```diff
@@ -45,7 +47,7 @@
 
   private String regionPath;
 
-  private Set<Object> entryKeys;
+  private Map<Object, Boolean> entryKeys;
```

Comment:
I think boolean may not be enough to handle the "read" locks case. You may need to add a counter to see how many "read" locks hold by existing transactions. When finishing a transaction, you release "read" lock by decrementing the counter, so that you do not lose other "read" locks by other concurrent transaction also holding the "read" lock.